File: worker/internal/wire/wire.go
// Package wire implements the DisputeVault DV1E/DV1V wire format
// (PROTOCOL.md §3), mirroring sdk/src/wire.ts exactly.
//
//	offset  size  field
//	0       4     magic     ASCII "DV1E" (evidence) | "DV1V" (verdict)
//	4       2     json_len  uint16 big-endian
//	6       n     json      canonical JSON (n == json_len)
//	6+n     ...   padding   0x00 up to the fixed wire size
//
// Rules enforced (identical to the SDK):
//   - json_len == actual JSON byte count; 6 + json_len <= wire size
//   - padding bytes MUST be zero
//   - the magic doubles as domain separation
//   - a wire that violates any rule is rejected, never repaired
package wire

import (
	"encoding/binary"
	"encoding/json"
	"unicode/utf8"

	"disputevault/worker/internal/canonical"
)

const (
	EvidenceMagic = "DV1E"
	VerdictMagic  = "DV1V"
)

// Kind selects the wire type; it fixes both the magic and the sizes.
type Kind int

const (
	Evidence Kind = iota
	Verdict
)

func (k Kind) String() string {
	if k == Verdict {
		return VerdictMagic
	}
	return EvidenceMagic
}

func (k Kind) wireSize() int {
	if k == Verdict {
		return canonical.VerdictWireSize
	}
	return canonical.EvidenceWireSize
}

func (k Kind) maxJSON() int {
	if k == Verdict {
		return canonical.VerdictMaxJSON
	}
	return canonical.EvidenceMaxJSON
}

func (k Kind) magic() string {
	if k == Verdict {
		return VerdictMagic
	}
	return EvidenceMagic
}

// Make frames a canonical JSON string into a fixed-size wire. The JSON byte
// count must fit the kind's budget; oversized payloads are rejected, never
// truncated.
func Make(kind Kind, canonicalJSON string) ([]byte, error) {
	jsonBytes := []byte(canonicalJSON) // Go strings are UTF-8; len counts BYTES
	if len(jsonBytes) > kind.maxJSON() {
		return nil, canonical.NewProtocolError(
			"canonical JSON is %d bytes; %s budget is %d (PROTOCOL §3) - rejected, never truncated",
			len(jsonBytes), kind, kind.maxJSON())
	}
	out := make([]byte, kind.wireSize())
	copy(out[0:4], kind.magic())
	binary.BigEndian.PutUint16(out[4:6], uint16(len(jsonBytes)))
	copy(out[6:], jsonBytes)
	return out, nil
}

// MakeEvidence frames a 256-byte DV1E wire.
func MakeEvidence(canonicalJSON string) ([]byte, error) { return Make(Evidence, canonicalJSON) }

// MakeVerdict frames a 192-byte DV1V wire.
func MakeVerdict(canonicalJSON string) ([]byte, error) { return Make(Verdict, canonicalJSON) }

// Parsed is the decoded view of a wire.
type Parsed struct {
	Kind    Kind
	Magic   string
	JSON    string
	JSONLen int
	Wire    []byte
}

// Parse strictly decodes a wire of the expected kind. Any deviation - wrong
// magic, inconsistent length, non-zero padding, invalid UTF-8, oversized
// JSON, unparsable JSON - is an error. Verifiers must use this before
// re-hashing.
func Parse(kind Kind, wire []byte) (*Parsed, error) {
	if len(wire) != kind.wireSize() {
		return nil, canonical.NewProtocolError(
			"%s wire must be exactly %d bytes, got %d", kind, kind.wireSize(), len(wire))
	}
	magic := string(wire[0:4])
	if magic != kind.magic() {
		return nil, canonical.NewProtocolError(
			"%s wire magic must be %q, got %q", kind, kind.magic(), magic)
	}
	jsonLen := int(binary.BigEndian.Uint16(wire[4:6]))
	if 6+jsonLen > kind.wireSize() {
		return nil, canonical.NewProtocolError(
			"%s wire json_len %d exceeds wire capacity (%d bytes)", kind, jsonLen, kind.wireSize())
	}
	if jsonLen > kind.maxJSON() {
		return nil, canonical.NewProtocolError(
			"%s wire json_len %d exceeds the %d-byte budget (PROTOCOL §3)", kind, jsonLen, kind.maxJSON())
	}
	jsonBytes := wire[6 : 6+jsonLen]
	if !utf8.Valid(jsonBytes) {
		return nil, canonical.NewProtocolError("%s wire JSON is not valid UTF-8", kind)
	}
	// json_len must equal the actual JSON byte count; a too-large length
	// field swallows padding bytes into the JSON, which cannot parse.
	if !json.Valid(jsonBytes) {
		return nil, canonical.NewProtocolError(
			"%s wire json_len %d is inconsistent with the embedded JSON (PROTOCOL §3)", kind, jsonLen)
	}
	for i := 6 + jsonLen; i < kind.wireSize(); i++ {
		if wire[i] != 0 {
			return nil, canonical.NewProtocolError(
				"%s wire has non-zero padding at offset %d (PROTOCOL §3)", kind, i)
		}
	}
	return &Parsed{
		Kind:    kind,
		Magic:   magic,
		JSON:    string(jsonBytes),
		JSONLen: jsonLen,
		Wire:    append([]byte(nil), wire...),
	}, nil
}

// ParseEvidence strictly decodes a DV1E wire.
func ParseEvidence(wire []byte) (*Parsed, error) { return Parse(Evidence, wire) }

// ParseVerdict strictly decodes a DV1V wire.
func ParseVerdict(wire []byte) (*Parsed, error) { return Parse(Verdict, wire) }

File: sdk/src/wire.ts
/**
 * DisputeVault wire format (PROTOCOL.md §3).
 *
 *   offset  size  field
 *   0       4     magic          ASCII "DV1E" (evidence) | "DV1V" (verdict)
 *   4       2     json_len       uint16 big-endian
 *   6       n     json           canonical JSON (n == json_len)
 *   6+n     ...   padding        0x00 up to the fixed wire size
 *
 * Rules enforced here:
 *   - json_len == actual JSON byte count; 6 + json_len <= wire size
 *   - padding bytes MUST be zero
 *   - magic doubles as domain separation (DV1E vs DV1V never interchangeable)
 *   - a wire that violates any rule is rejected, never repaired
 */
import { ProtocolError, byteLength } from './canonical.js';
import { WIRE } from './schema.js';

export const EVIDENCE_MAGIC = 'DV1E';
export const VERDICT_MAGIC = 'DV1V';

const encoder = new TextEncoder();
const decoder = new TextDecoder('utf-8', { fatal: true });

export type WireKind = 'evidence' | 'verdict';

const WIRE_SIZE: Record<WireKind, number> = {
  evidence: WIRE.evidenceWireSize,
  verdict: WIRE.verdictWireSize,
};

const MAX_JSON: Record<WireKind, number> = {
  evidence: WIRE.evidenceMaxJsonBytes,
  verdict: WIRE.verdictMaxJsonBytes,
};

const MAGIC: Record<WireKind, string> = {
  evidence: EVIDENCE_MAGIC,
  verdict: VERDICT_MAGIC,
};

/** Frame a canonical JSON string into a fixed-size DV1E/DV1V wire. */
export function makeWire(kind: WireKind, canonicalJson: string): Uint8Array {
  const jsonBytes = encoder.encode(canonicalJson);
  const wireSize = WIRE_SIZE[kind];
  const maxJson = MAX_JSON[kind];
  if (jsonBytes.length > maxJson) {
    throw new ProtocolError(
      `canonical JSON is ${jsonBytes.length} bytes; ${kind} budget is ${maxJson} (PROTOCOL §3) - rejected, never truncated`,
    );
  }
  const wire = new Uint8Array(wireSize);
  wire.set(encoder.encode(MAGIC[kind]), 0);
  wire[4] = (jsonBytes.length >> 8) & 0xff;
  wire[5] = jsonBytes.length & 0xff;
  wire.set(jsonBytes, 6);
  return wire;
}

export const makeEvidenceWire = (canonicalJson: string): Uint8Array =>
  makeWire('evidence', canonicalJson);
export const makeVerdictWire = (canonicalJson: string): Uint8Array =>
  makeWire('verdict', canonicalJson);

/** Parsed view of a wire. */
export interface ParsedWire {
  kind: WireKind;
  magic: string;
  json: string;
  jsonBytes: Uint8Array;
  jsonLen: number;
  wire: Uint8Array;
}

/**
 * Strictly parse a wire of the expected kind. Any deviation - wrong magic,
 * inconsistent length field, non-zero padding, invalid UTF-8, oversized JSON
 * - throws ProtocolError. Verifiers MUST use this before re-hashing.
 */
export function parseWire(kind: WireKind, wire: Uint8Array): ParsedWire {
  const wireSize = WIRE_SIZE[kind];
  if (wire.length !== wireSize) {
    throw new ProtocolError(
      `${kind} wire must be exactly ${wireSize} bytes, got ${wire.length}`,
    );
  }
  const magic = decoder.decode(wire.subarray(0, 4));
  if (magic !== MAGIC[kind]) {
    throw new ProtocolError(
      `${kind} wire magic must be "${MAGIC[kind]}", got "${magic}"`,
    );
  }
  const jsonLen = (wire[4]! << 8) | wire[5]!;
  if (6 + jsonLen > wireSize) {
    throw new ProtocolError(
      `${kind} wire json_len ${jsonLen} exceeds wire capacity (${wireSize} bytes)`,
    );
  }
  if (jsonLen > MAX_JSON[kind]) {
    throw new ProtocolError(
      `${kind} wire json_len ${jsonLen} exceeds the ${MAX_JSON[kind]}-byte budget (PROTOCOL §3)`,
    );
  }
  const jsonBytes = wire.slice(6, 6 + jsonLen);
  const json = decoder.decode(jsonBytes); // fatal: throws on invalid UTF-8
  // json_len must equal the actual JSON byte count; a too-large length field
  // swallows padding bytes into the JSON, which cannot parse as JSON.
  try {
    JSON.parse(json);
  } catch {
    throw new ProtocolError(
      `${kind} wire json_len ${jsonLen} is inconsistent with the embedded JSON (PROTOCOL §3)`,
    );
  }
  for (let i = 6 + jsonLen; i < wireSize; i++) {
    if (wire[i] !== 0) {
      throw new ProtocolError(
        `${kind} wire has non-zero padding at offset ${i} (PROTOCOL §3)`,
      );
    }
  }
  return { kind, magic, json, jsonBytes, jsonLen, wire: wire.slice() };
}
